### ISSUE TITLE
Aysnc Select retrigger loadOptions?

### ISSUE THREAD

**Issue by AaronMcCloskey** [JedWatson/react-select#4118]
I have an async `react-select` component

```
import React, { useRef, useState } from "react";
import Async from "react-select/async";

const MySelect = ({
  label,
  options,
  asyncSelect,
  loadOptions,
  components,
  placeholder,
  fixedHeight,
  onSelect,
  ...props
}) => {
  const ref = useRef();
  const asyncRef = useRef();

  const [fieldValue, setFieldValue] = useState();

  // Gets called when an option is chosen or created.
  const onChange = value => {
    setFieldValue(value);
    if (onSelect) onSelect(value);
  };

  // Toggle menu by presence of input.
  // Reset value if input is emptied.
  const onInputChange = (value, context) => {
    if (!value && context.action === "input-change") {
      onChange(null);
    }
  };

  // Restore input value.
  const onFocus = () => {
    if (fieldValue) {
      asyncRef.current.state.inputValue = fieldValue.label;
      asyncRef.current.select.state.inputValue = fieldValue.label;
      asyncRef.current.handleInputChange(fieldValue.label);
    }
  };

  // Close menu when pressing enter. Submit does not work on mobile.
  const onKeyDown = event => {
    if (event.keyCode === 13) {
      onMenuClose();
    }
  };

  // Blur select element to trigger onFocus on next click.
  const onMenuClose = () => {
    asyncRef.current.select.blur();
  };

  return (
    <>
      <div ref={ref} />
      <Async
        ref={asyncRef}
        {...props}
        value={fieldValue}
        components={components ? components : null}
        loadOptions={loadOptions ? loadOptions : () => null}
        placeholder={placeholder ? placeholder : "Select..."}
        searchPromptText="Start Typing"
        autoload
        isSearchable
        searchable
        cacheOptions
        // Hook into events to make the editing work.
        onChange={onChange}
        onInputChange={onInputChange}
        onFocus={onFocus}
        onKeyDown={onKeyDown}
        onMenuClose={onMenuClose}
      />
    </>
  );
};

export default MySelect;
```

I use this component in another component to load an array and filter options from the input value

```
const loadOptions = async inputValue => {
  return new Promise(resolve => resolve(getOptions(inputValue)));
};

const getOptions = async inputValue => {
  // how to trigger this to be called within loadOptions when option is selected?
  if (asyncOptions && asyncOptions.length > 0) {
    const options = asyncOptions.filter(item =>
      item.label.toLowerCase().startsWith(inputValue.trim().toLowerCase())
    );
    if (options && options.length > 0) {
      return options;
    }
    return [];
  }
  await delayWithPromise(1000);
  return [];
};

return (
  <div className="App">
    <MySelect
      name="addressLookup"
      className="addressLookupContainer"
      label="Address Lookup"
      asyncSelect
      components={{
        DropdownIndicator: () => null,
        IndicatorSeparator: () => null,
        NoOptionsMessage: val => <NoOptionsMsg>No Options</NoOptionsMsg>
      }}
      loadOptions={loadOptions}
      onSelect={val => console.log(`${val.value} selected`)}
    />
  </div>
);
```

However, I want it so when I select an option, it then loads another list (without the list closing, instead loads a new list in the old lists place), but I am unsure how I can retrigger the `loadOptions` prop within `react-select/async` when I click on an option?

An example would be an address search, where you search for a street/ post/zip code and there is an option stating how many address with that street, then selecting that option would then trigger `loadOptions` to make an API call to find all address in that street, therefore, bringing in a new list of options. Hopefully, that makes sense?

Any help would be greatly appreciated.

I have a [CodeSandbox][1] with an example.

  [1]: https://codesandbox.io/s/solitary-brook-bjcni

**Comment by ebonow** [comment]
@AaronMcCloskey I believe I understand what you are trying to accomplish. My first stab at this would be to ensure that your loadOptions function would need to be able to know whether there has been a selection.

```
const [fieldValue, setFieldValue] = useState();
  
const loadOptions = async (inputValue, callback) => {
  const stringCompare = (item) => item.label.toLowerCase().startsWith(inputValue.trim().toLowercase());

  const options = await !fieldValue ? getStreets() : getStreetAddresses();
  callback(options.filter(stringCompare))
}
```

In regards to retriggering it, you could blur and refocus to get the menu back open onChange. 

I am not sure I fully understand, so if there is something missing in understanding your problem, please let me know.

**Comment by AaronMcCloskey** [comment]
@ebonow Yeah, it's mainly the retriggering I am unsure of. Figuring out what is selected isn't an issue, in my actual code, I already have that part sorted. (I have updated the CodeSandbox to reflect this)

I am essential wanting a user to select the first option, then, the other list loads, without the list closing, or at least preventing the user from having to click the input again to get the new list, as it's not very intuitive. Even if it closes and then reopens I'm fine with, it's just that additional click which isn't instantly obvious that the user could view the further options.

Hopefully, that clears it up a little?

**Comment by ebonow** [comment]
@AaronMcCloskey Spent some time thinking on this with some approaches in mind. First the solution I was able to get working though admittedly quick and dirty. Add the `openMenuOnFocus` prop and then modify the onChange

``` 
// MySelect.js
  const onChange = value => {
    setFieldValue(value);

    if (onSelect) {
      onSelect(value);
      asyncRef.current.select.blur();
      setTimeout(() => asyncRef.current.select.focus(), 1);
    }
  };
```

Not ideal, but it works. You would likely want to abstract this into the onSelect method call but you would need to forward the refs so you have access to the blur and focus methods for the MySelect from your App.

**Comment by ebonow** [comment]
Some minor things, but it appears you are still using some removed or renamed props [(autoload and searchable)](https://react-select.com/upgrade-guide#prop-update-guide) Also, isSearchable is true by default.

Also your Sandbox example was made more difficult by the cacheOptions props as it would ignore the new options "chunk" so you may need to disable it for this use case.

**Comment by ebonow** [comment]
As for other approaches I tried, I thought the most straight forward way of approaching this is passing in the prop `closeMenuOnSelect={!!firstValue}`. This works for keeping the menu open, but menu did not update as you say. 

There was some state behavior I was unable to figure out. I tried calling loadOptions from MySelect onChange method. Even after putting it in a timeout, my console logs were telling me that the firstValue was not set. 

I use a similar nested pattern to what it seems like you are doing here, but I am instead using `getDefaultOptions` to filter down my list based on the user's criteria and using `menuIsOpen` to keep the menu open. I am also using the callback method instead of promises for loadOptions, so that could potentially be one other difference.

<img width="347" alt="Screen Shot 2020-07-09 at 10 41 10 PM" src="https://user-images.githubusercontent.com/1007579/87229478-dd4ec500-c36d-11ea-9405-aba83954f233.png">
